### ISSUE TITLE
PlantUML: Different behaviour whether LATEX_OUTPUT has a final slash or not.

### ISSUE THREAD

**Issue by wosrediinanatour** [doxygen/doxygen#7558]
I have used latest version (commit).

I got following error with `LATEX_OUTPUT           = /opt/app-root/src/build/doc`: 
```
Generating docs for page md_build_doc_README_modified...
/opt/app-root/src/build/doc/README_modified.md:62: warning: Unsupported xml/html tag <hostname> found
*** instance() : m_cachedPlantumlAllContent = []
*** writePlantUMLSource fileName: ApplicationDetails
*** writePlantUMLSource outDir: /opt/app-root/src/build/doc
*** writePlantUMLSource baseName: /opt/app-root/src/build/doc/ApplicationDetails
*** writePlantUMLSource puName: ApplicationDetails
*** writePlantUMLSource imgName: ApplicationDetails.eps
*** writePlantUMLSource generateType: doc
*** PlantumlManager::insert key:doc ,value:ApplicationDetails
*** PlantumlManager::addPlantumlContent find: -1
*** PlantumlManager::print Files PlantumlFiles key:doc size:1
*** PlantumlManager::print             list:ApplicationDetails
*** PlantumlManager::print Content PlantumlContent key:doc
*** PlantumlManager::print                 Content :@startuml ApplicationDetails.eps
!include ApplicationDetails.puml
@enduml

*** writePlantUMLSource generateType: doc
*** writePlantUMLSource fileName: Tableconfiguration
*** writePlantUMLSource outDir: /opt/app-root/src/build/doc
*** writePlantUMLSource baseName: /opt/app-root/src/build/doc/Tableconfiguration
*** writePlantUMLSource puName: Tableconfiguration
*** writePlantUMLSource imgName: Tableconfiguration.eps
*** writePlantUMLSource generateType: doc
*** PlantumlManager::insert key:doc ,value:Tableconfiguration
*** PlantumlManager::addPlantumlContent find: -1
*** PlantumlManager::print Files PlantumlFiles key:doc size:2
*** PlantumlManager::print             list:ApplicationDetails
*** PlantumlManager::print             list:Tableconfiguration
*** PlantumlManager::print Content PlantumlContent key:doc
*** PlantumlManager::print                 Content :@startuml ApplicationDetails.eps
!include ApplicationDetails.puml
@enduml
@startuml Tableconfiguration.eps
!include Tableconfiguration.puml
@enduml

*** writePlantUMLSource generateType: doc


...

Running plantuml with JAVA...
*** PlantumlManager::run
Generating PlantUML eps Files in doc
*** PlantumlManager::runPlantumlContent Running Plantuml arguments:-Dplantuml.include.path="doc" -Djava.awt.headless=true -jar "/usr/share/java/plantuml.jar" -o "/opt/app-root/src/build/doc/doc" -charset UTF-8 -teps "/opt/app-root/src/bu
ild/doc/doc/inline_umlgraph_epsdoc.pu"
Error line 1 in file: /opt/app-root/src/build/doc/doc/inline_umlgraph_epsdoc.pu
Some diagram description contains errors
error: Problems running PlantUML. Verify that the command 'java -jar "/usr/share/java/plantuml.jar" -h' works from the command line. Exit code: 200
*** PlantumlManager::runPlantumlContent Running epstopdf
```

With
`LATEX_OUTPUT           = /opt/app-root/src/build/doc/`
it works and I get following (even with message `QGDict::hashAsciiKey: Invalid null key`):
```
Generating page documentation...
Generating docs for page md_build_doc_README_modified...
/opt/app-root/src/build/doc/README_modified.md:62: warning: Unsupported xml/html tag <hostname> found
*** instance() : m_cachedPlantumlAllContent = []
*** writePlantUMLSource fileName: ApplicationDetails
*** writePlantUMLSource outDir: /opt/app-root/src/build/doc/
*** writePlantUMLSource baseName: /opt/app-root/src/build/doc//ApplicationDetails
*** writePlantUMLSource puName: ApplicationDetails
*** writePlantUMLSource imgName: ApplicationDetails.eps
*** writePlantUMLSource generateType:
*** PlantumlManager::insert key: ,value:ApplicationDetails
*** PlantumlManager::addPlantumlContent find: -1
QGDict::hashAsciiKey: Invalid null key
QGDict::hashAsciiKey: Invalid null key
*** PlantumlManager::print Files PlantumlFiles key: size:1
*** PlantumlManager::print             list:ApplicationDetails
QGDict::hashAsciiKey: Invalid null key
QGDict::hashAsciiKey: Invalid null key
*** PlantumlManager::print Content PlantumlContent key:
*** PlantumlManager::print                 Content :@startuml ApplicationDetails.eps
!include ApplicationDetails.puml
@enduml

*** writePlantUMLSource generateType:
*** writePlantUMLSource fileName: Tableconfiguration
*** writePlantUMLSource outDir: /opt/app-root/src/build/doc/
*** writePlantUMLSource baseName: /opt/app-root/src/build/doc//Tableconfiguration
*** writePlantUMLSource puName: Tableconfiguration
*** writePlantUMLSource imgName: Tableconfiguration.eps
*** writePlantUMLSource generateType:
*** PlantumlManager::insert key: ,value:Tableconfiguration
*** PlantumlManager::addPlantumlContent find: -1
QGDict::hashAsciiKey: Invalid null key
*** PlantumlManager::print Files PlantumlFiles key: size:2
*** PlantumlManager::print             list:ApplicationDetails
*** PlantumlManager::print             list:Tableconfiguration
QGDict::hashAsciiKey: Invalid null key
*** PlantumlManager::print Content PlantumlContent key:
*** PlantumlManager::print                 Content :@startuml ApplicationDetails.eps
!include ApplicationDetails.puml
@enduml
@startuml Tableconfiguration.eps
!include Tableconfiguration.puml
@enduml

...

Running plantuml with JAVA...
*** PlantumlManager::run
Generating PlantUML eps Files in
*** PlantumlManager::runPlantumlContent Running Plantuml arguments:-Dplantuml.include.path="doc" -Djava.awt.headless=true -jar "/usr/share/java/plantuml.jar" -o "/opt/app-root/src/build/doc/" -charset UTF-8 -teps "/opt/app-root/src/build
/doc//inline_umlgraph_eps.pu"
*** PlantumlManager::runPlantumlContent Remove /opt/app-root/src/build/doc//inline_umlgraph_eps.pu file
*** PlantumlManager::runPlantumlContent Running epstopdf
QGDict::hashAsciiKey: Invalid null key
Running dot...

```

The difference is the *second* sub-folder `doc` in `*** PlantumlManager::runPlantumlContent Running Plantuml arguments:-Dplantuml.include.path="doc" -Djava.awt.headless=true -jar "/usr/share/java/plantuml.jar" -o "/opt/app-root/src/build/doc/doc" -charset UTF-8 -teps "/opt/app-root/src/bu
ild/doc/doc/inline_umlgraph_epsdoc.pu"`. This second sub-folder **does not exist**. 

Line 75 of File `src/plantuml.cpp` works different for with and without final slash: 
```
  QCString qcOutDir(outDir);
  uint pos = qcOutDir.findRev("/");
  QCString generateType(qcOutDir.right(qcOutDir.length() - (pos + 1)) );
  Debug::print(Debug::Plantuml,0,"*** %s generateType: %s\n","writePlantUMLSource",qPrint(generateType));
  PlantumlManager::instance()->insert(generateType,puName,format,text);
```

**Comment by albert-github** [comment]
Looks clear but (to be sure we are talking about the same):
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem?  Please don't add external links as they might not be persistent.

As a small side note:
You wrote "I have used latest version (commit)." which is good, would even be better when you write the commit ID with it. when running your own compiled doxygen version like `doxygen -v` you wil lget a line like

> 1.8.18 (59fc1a94e84278dedc80569b72756f336a705693)

using this in "I have used latest version (commit)." makes it much easier to find what the used version was.

**Comment by albert-github** [comment]
I'm not able to reproduce the problem (I can see the `QGDict::hashAsciiKey: Invalid null key`, but this is of no influence, but should not appear and is worth another issue).

I think it might have multiple reasons to be able to see what is going wrong it is needed to have:
- the doxygen configuration file **with** the slash
- the doxygen configuration file **without** the slash
- the used project (problem might be in an included plantuml file)


Note: when running a file `inline_umlgraph_cache_all.pu` is generated so when rerunning for the tests please remove this `inline_umlgraph_cache_all.pu` file and also the `html` and `latex` output directories.

**Comment by albert-github** [comment]
Regarding the `QGDict::hashAsciiKey: Invalid null key` I've just pushed a proposed patch, pull request #7559

**Comment by albert-github** [comment]
Regarding a problem with a Windows output path I've just pushed a proposed patch, pull request #7560

**Comment by wosrediinanatour** [comment]
> Looks clear but (to be sure we are talking about the same):
> 
>     * Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem?  Please don't add external links as they might not be persistent.
Here you are:

 * I made a minimal example: [example.tar.gz](https://github.com/doxygen/doxygen/files/4143183/example.tar.gz)
 * Versions:
   * Doxygen: 1.8.18 (348f0af54ed1cc57eb1df62d47d6d869bc769e1d*).
   * PlantUML: (0.000 - 48 Mo) 42 Mo - PlantUML Version 1.2019.06
* Here is my output: [output.txt](https://github.com/doxygen/doxygen/files/4143188/output.txt)
 * For testing, please 
    * untar into Folder `doxygen`
    * search & replace following string in Doxyfile: `/opt/app-root/src/` by your absolute path.
    * I tried to change it to an relative path, but got different results. Maybe that is important to note. I think that you can replace it with yours.
    * `doxygen -d plantuml`

Hope it helps!

**Comment by albert-github** [comment]
I have been able to reproduce the problem (wasn't easy as for Cygwin there is no java available, as far as I can see; for other java programs I use some horrible and dedicated path conversions and I have to use #7560 as well).
The entire setup is a bit strange as `LATEX_OUTPUT` and `OUTPUT_DIRECTORY` are the same directory and this directory is also used as `INPUT`.
But still I don't think doxygen should have a problem with it.

Note: in the example.tar.gz the directory structure is gone (so by default there are no `doc` and `doxygen` directories, all files are in the "root" where it is extracted).

**Comment by wosrediinanatour** [comment]
Regarding example: yes, it occurs even in this flat case without directory structure! (good, since the original system is far more complex!)

The main Problem is the additional subdirectory, which does not exist. So no file can be created there. So, I think, that you don't need Java/PlantUML for debugging. 
I was puzzled about the code snipped I posted earlier - I don't understand the reason behind it. What is it supposed to do? An older Doxygen, without this code section and others, is working, well.

Maybe I can provide more info next week.

El Domingo 2 de febrero de 2020, albert-github escribió:
> I have been able to reproduce the problem (wasn't easy as for Cygwin there is no java available, as far as I can see; for other java programs I use some horrible and dedicated path conversions and I have to use #7560 as well).
> The entire setup is a bit strange as `LATEX_OUTPUT` and `OUTPUT_DIRECTORY` are the same directory and this directory is also used as `INPUT`.
> But still I don't think doxygen should have a problem with it.
> 
> Note: in the example.tar.gz the directory structure is gone (so by default there are no `doc` and `doxygen` directories, all files are in the "root" where it is extracted).
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/doxygen/doxygen/issues/7558#issuecomment-58114159

-- 
Enviado desde mi dispositivo Sailfish

**Comment by albert-github** [comment]
I have already the possibility to reproduce the problem and a reasonable clear picture of what is going wrong, so no need for more info at the moment. It looks like the fact that the absolute output directories for `LATEX_OUTPUT` etc. are not handled properly.

**Comment by wosrediinanatour** [comment]
oh...I missed the most important part of your previous message.

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #7569, this handles the main problem (i.e. the non relative path), other issues were handled in the proposed pull requests #7559 and #7560.

**Comment by albert-github** [comment]
Code #7569 has been integrated in master on github (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.18.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
